ktor-http/common/src/io/ktor/http/HttpHeaderValueParser.kt

```diff
@@ -174,27 +178,43 @@ private fun parseHeaderValueParameterValue(value: String, start: Int): Pair<Int,
 }
 
 private fun parseHeaderValueParameterValueQuoted(value: String, start: Int): Pair<Int, String> {
-    var pos = start
-    val sb = StringBuilder()
-    while (pos <= value.lastIndex) {
-        val current = value[pos]
-        when (current) {
-            '"' -> return pos + 1 to sb.toString()
-            '\\' -> {
-                if (pos < value.lastIndex - 2) {
-                    sb.append(value[pos + 1])
-                    pos += 2
-                } // quoted value
-                else {
-                    sb.append(current)
-                    pos++ // broken value, escape at the end
-                }
+    var position = start
+    val builder = StringBuilder()
+    loop@ while (position <= value.lastIndex) {
+        val current = value[position]
+
+        when {
+            current == '"' && value.nextIsSemicolonOrEnd(position) -> {
+                return position + 1 to builder.toString()
             }
-            else -> {
-                sb.append(current)
-                pos++
+            current == '\\' && position < value.lastIndex - 2 -> {
+                builder.append(value[position + 1])
+                position += 2
+                continue@loop
+            }
+        }
+
+        builder.append(current)
+        position++
+    }
+
+    // The value is unquoted here
+    return position to '"' + builder.toString()
+}
+
+private fun String.nextIsSemicolonOrEnd(start: Int): Boolean {
+    var position = start + 1
+    loop@ while (position < length) {
+        when (get(position)) {
+            ';' -> return true
+            ' ' -> {
+                position += 1
+                continue@loop
```

Comment:
Nit: 
`while (position < length && get(position) == ' ') position++`
`return get(position) == ';'`

probably would be easier to read
But feel free to leave it as it is.
LGTM

Reply:
Yep, I actually like your way. Except for missing `{}` in while loop

Reply:
`parseHeaderValueParameterValueQuoted` performs the check and append `"` at the start if failed to find a closing one

Reply:
Yep, sounds good